### ISSUE TITLE
[Epic] Compute instance isolation

### ISSUE THREAD

**Issue by frankmcsherry** [MaterializeInc/materialize#11242]
`CREATE CLUSTER` is backed by COMPUTE's `CreateInstance` command, which instantiates new cluster instances (timely dataflow instances).

### Goals

* Customers can create multiple cluster instances that are isolated from each other (separate fault domain)
* Clusters are isolated from ADAPTER (user sessions, coordinator, controller, etc.)
* Clusters consume data from STORAGE (see also #11434)

### Non-Goals

* Sizing of cluster instance(s)

### Acceptance criteria

Users can create independent clusters and interact with them separately. Errors within one cluster do not affect other clusters. Specifically, other clusters remain responsive, can ingest new data, and produce new results.

Errors constitute situations where Materialize runs out of resources (CPU, RAM), or suffers from an internal or external fault.

### Subtasks

- [x] #11507 (owned by ADAPTER)
- [x] #12220
- [x] #12251 (owned by ADAPTER)
- [x] #12255

### Testing
- [x] #12220

### Bugs
- [ ] #12403
- [x] #12294
- [x] #12251
- [x] #12549 (operations against replicas of one cluster affect result set correctness in another cluster)
- [x] #12546 (a bit of a feature request; theoretically related)
- [ ] #12595 (coord crashes because storaged has crashed)
- [ ] #12885 once one replica is brought down, the second one will crash as well.
- [ ] #12951

### Related

* [design/20220214_create_cluster.md](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/20220214_create_cluster.md)

### Open Questions

* Isolated workflow UX good enough?
  * Cancellation behavior when cluster "unhealthy"?

**Comment by sploiselle** [comment]
tl;dr `ADAPTER` and `COMPUTE` seem sufficiently isolated, but we could make it a little more obvious.

# ADAPTER-COMPUTE isolation

The ADAPTER works as the intermediary between clients and COMPUTE + STORAGE.

Elsewhere, we have identified that ADAPTER is not currently built to be isolated from STORAGE failures, and this epic is solely focused on COMPUTE isolation.

This provides us with three areas to examine w/r/t ADAPTER-COMPUTE isolation:

- ADAPTER to COMPUTE
- Session to ADAPTER
- Intra-ADAPTER

## ADAPTER to COMPUTE (ComputeController)

The adapter sends client and system commands to "compute instances" (aka clusters in SQL) via the coordinator's dataflow client (`self.dataflow_client.compute_mut(compute_instance).unwrap()`, which returns a `ComputeController`, whose methods then provide the API to use clusters).

A number of the `ComputeController`'s methods are `async`, which means we have a number of spots in the ADAPTER that `await` a response from the `ComputeController`--to guarantee isolation between the ADAPTER and COMPUTE, it seems like we need to institute some kind of timeout/failover mechanism to ensure that ADAPTER doesn't wait forever for COMPUTE to do some work on a cluster with an arbitrary number of replicas, some of which might be down.

However, this relationship is not quite as it seems in the ADAPTER. To communicate with replicas the `ComputeController`'s state has a client known as `ActiveReplication`--commands that affect compute instances are propagated to all replicas (in the general case) using this client. To communicate with all replicas, `ActiveRepliation` implements mplements `GenericClient`, which requires `async fn send(...`--and that is actually where the async calls originate from. `ActiveReplication::send`, though, does not require any async code--`ActiveReplication` sends commands to replicas without awaiting their replies.

This means that all calls to `compute_mut()...await` are not truly asynchronous and do not impose any actual awaits on the coordinator. This obviates the need to impose any kind of strict isolation mechanism between the ADAPTER and COMPUTE, or rather, the isolation mechanism already exists, despite the difficulty in seeing it.

### Why are we awaiting?

Right now the `ComputeController` also bundles a `StorageController`, which we rely on for e.g. controlling read policies. The `StorageController` relies on a TCP connection to the storage service, so it can only send asynchronously.

Given the dependency between COMPUTE and STORAGE, it's unlikely that the `ComputeController` will divest itself of this TCP connection, meaning the `ComputeController` is likely to always present an `async` API to the ADAPTER (again, despite the fact that the actual COMPUTE resources don't need to be handled asynchronously).

### Action item

Given that there's nothing we need to handle here, technically, we can simply attempt to make the existing isolation strategy more legible in the coordinator:
- Liberally add documentation to `coord.rs`
- Put some safeguards around `ActiveReplation::send` to ensure it doesn't inherit any `async` methods, e.g. creating a private synchronous function that powers the asynchronous `ActiveReplation::send`

## Session to adapter

Sessions send statements to the ADAPTER, which forward them to COMPUTE, and then receives responses from COMPUTE and then forwards them to the session. Isolation in this realm requires that the ADAPTER continues to work irrespective of the session's state--because this is largely a matter of handing COMPUTE responses back to the session, this is what we focus on.

### COMPUTE responses

The general architecture of the coordinator is to send commands along the `ComputeController`, and then regularly check for `ControllerResponse`s.

The ADAPTER remains isolated from COMPUTE in the following scenarios with the following strategies:

#### No response

**Peeks + tails**
If COMPUTE will simply never respond, Materialize will not send a response to the client, and will depend on the client to eventually terminate the session. This lingering session is a nominal drain on the ADAPTER resources, so isn't something we've actively addressing. However, we could rely on some timeout mechanism to proactively terminate sessions we believe will never receive a response. This is what PG's strategy is, and is something we could implement.

Alternatively, Jan's work to detect unhealthy replicas could signal to us that we need to terminate all connections to a cluster if all replicas are unhealthy (e.g. if the cluster is servicing a crash-loop inducing statement).

#### Duplicate responses

**Peeks**
Peek responses come into the coordinator with a UUID attached. If we get a duplicate response (i.e. two responses with the same UUID), the first response will remove the UUID from pending peeks, and the second response will simply get dropped on the floor.

**Tails**
Each tail's identified by a `GlobalId` and we use the same mechanism of handling duplicate responses by dropping the second on the floor.

This behaved differently in previous version of Materialize, and was rectified in https://github.com/MaterializeInc/materialize/pull/12427/commits/e2eb90c83456364e74d12ed5bf57c62b490585f4

#### Response with no recipient

**Peeks**
e.g. peek cancellation is best effort, and we make no guarantee that the coordinator will not receive a response for a peek that it cancelled. We treat this similarly to the duplicate response case, but the client/session cancellation/termination is equivalent to the first response (i.e. removes the pending peek).

**Tails**
Tails are more closely tied to sessions (i.e. each session keeps track of its own pending tails), so if a session drops, so too does its tail. Any responses that the `TAIL` is ready are summarily ignored. This is essentially equivalent to how we handle duplicate responses.

This behaved differently in previous version of Materialize, and was rectified in https://github.com/MaterializeInc/materialize/pull/12427/commits/e2eb90c83456364e74d12ed5bf57c62b490585f4 + https://github.com/MaterializeInc/materialize/pull/12683

## Intra-ADAPTER

Components of the ADAPTER can depend on COMPUTE behavior, where the ADAPTER itself behaves as a client; so we need to ensure client-like behavior of the ADAPTER in these places.

Currently, the only place where the ADAPTER itself behaves like a client to COMPUTE in `ReadThenWrite` operations.

### Table locks

The ADAPTER has the option to lock all tables, preventing any writes to tables. It holds this lock during the sequencing of `ReadThenWrite` operations, e.g. UPDATE, as well as during the commit phase of any transaction.

Phillip discovered that if we held the lock during a `ReadThenWrite` and waited for a downed cluster to reply, we would hold the lock indefinitely, preventing all future until the client cancelled the statement.

There was a long discussion about approaches to mitigate this (#12251) and we settled on using a configurable timeout for the time being. This essentially brings our expected session behavior into the ADAPTER (if something isn't responding, we expect the client to terminate the session).

We are confident that there is nowhere else in the code affected by this because of the small surface areas where locks get held.

The current behavior was implemented in https://github.com/MaterializeInc/materialize/pull/12427

Note that there are other design options in this space that would let us remove the need to hold a pessimistic lock, such as using `TAIL` instead of `Peek` in `ReadThenWrite` operations. However, moving toward something like optimistic locks is more complex an requires a substantial amount of work to e.g. introduce a notion of priority to handle lock contention. This is purely an optimization and not something worth taking on until we hear from users that the configurable `ReadThenWrite` timeout is unacceptable, or we decided to invest more fully in improving the ergonomics/performance of tables in general.

# Summary

The only real work item I was able to identify in this analysis is adding safeguards to ensure `ActiveReplication::send` doesn't become a truly asynchronous function. Generally, I believe the ADAPTER is largely isolated from COMPUTE. Surely there are elements I've overlooked, but every avenue of investigation I pursued had a satisfactory answer.

**Comment by uce** [comment]
I'm closing this issue as the main subtasks have been completed.
